napari/utils/chunk/_request.py

```diff
@@ -0,0 +1,309 @@
+
+import numpy as np
+
+from ...types import ArrayLike, Dict
+from ...utils.perf import perf_counter_ns, PerfEvent, timers
+from ._util import get_data_id
+
+LOGGER = logging.getLogger("ChunkLoader")
+
+# We convert slices to tuple for hashing.
+SliceTuple = Tuple[int, int, int]
+
+LayerData = Union[ArrayLike, List[ArrayLike]]
+
+
+class TimeSpan:
+    """A span of time in seconds.
+
+    Parameters
+    ----------
+    start_seconds : float
+        Start of the time span.
+    end_seconds : float
+        End of the time span.
+    """
+
+    def __init__(self, start_seconds: float, end_seconds: float, args: dict):
+        self.start_seconds = start_seconds
+        self.end_seconds = end_seconds
+        self.args = args
+
+    @property
+    def duration_seconds(self) -> float:
+        """Return duration of the time span.
+
+        Returns
+        -------
+        float
+            Duration of the time span in seconds.
+        """
+        return self.end_seconds - self.start_seconds
+
+    @property
+    def duration_ms(self) -> float:
+        """Return duration of the time span.
+
+        Returns
+        -------
+        float
+            Duration of the time span in milliseconds.
+        """
+        return self.duration_seconds * 1000
+
+
+class ChunkKey:
+    """The key which a ChunkRequest will load.
+
+    Parameters
+    ----------
+    layer : Layer
+        The layer to load data for.
+    indices : ?
+        The indices to load from the layer
+    """
+
+    def __init__(self, layer, indices):
+        self.layer_id = id(layer)
+        self.data_id = get_data_id(layer)
+        self.data_level = layer._data_level
+
+        # Slice objects are not hashable, so turn them into tuples.
+        self.indices = tuple(_index_to_tuple(x) for x in indices)
+
+        # All together as one tuple for easy comparison.
+        self.key = (self.layer_id, self.data_id, self.data_level, self.indices)
+
+    def __str__(self):
+        return (
+            f"layer_id={self.layer_id} data_id={self.data_id} "
+            f"data_level={self.data_level} indices={self.indices}"
+        )
+
+    def __eq__(self, other):
+        return self.key == other.key
+
+
+class ChunkRequest:
+    """A request asking the ChunkLoader to load one or more arrays.
+
+    Parameters
+    ----------
+    layer_id : int
+        Python id() for the Layer requesting the chunk.
+    data_id : int
+        Python id() for the Layer._data requesting the chunk.
+    indices
+        The tuple of slices index into the data.
+    array : ArrayLike
+        Load the data from this array.
+
+    Attributes
+    ----------
+    layer_ref : weakref
+        Reference to the layer that submitted the request.
+    data_id : int
+        Python id() of the data in the layer.
+    load_seconds : float
+        Delay for this long during the load portion.
+    """
+
+    def __init__(self, key: ChunkKey, chunks: Dict[str, ArrayLike]):
```

Comment:
It happens that today `Image` only uses those 2 keys, but the idea is the `ChunkLoader` allows people to ask for any number of chunks with any names/keys, it doesn't and shouldn't care. For example with the octree/tiles it might give the some sort of row/column index as part of the chunk name. Or it might ask for 4 or 16 chunks at once if they are tiny.

So I think how it is expresses the intent. That the `ChunkLoader` doesn't care.

Reply:
In that case... I think it's strange to have

```python
    @property
    def image(self): ...
    
    #and
    @property
    def thumbnail_source(self): ...
```

as mentioned in another comment, I i feel like there's generally excessive `@property` use on the classes in this PR.  If you're trying to create a chunk loading API that just happens to use `image` and `thumbnail_source`, but could eventually use any key, then I think you should have a `ChunkRequest.get` method that passes the key to `ChunkRequest.get`... and then use that API directly elsewhere in our code